src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1547,3 +2101,367 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
 
 	return true;
 }
+
+
+#include "nodes/print.h"
+
+/*
+ * ReorderInsertSelectTargetListsIfExists reorders the target lists of INSERT/SELECT
+ * query which is required for deparsing purposes. The reordered query is returned.
+ *
+ * The necessity for this function comes from the fact that ruleutils.c is not supposed to be
+ * used on "rewritten" queries (i.e. ones that have been passed through QueryRewrite()).
+ * Query rewriting is the process in which views and such are expanded,
+ * and, INSERT/UPDATE targetlists are reordered to match the physical order,
+ * defaults etc. For the details of reordeing, see transformInsertRow().
+ */
+Query *
+ReorderInsertSelectTargetListsIfExists(Query *originalQuery)
+{
+	RangeTblEntry *insertRte = NULL;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+	ListCell *insertTargetEntryCell;
+	List *newSubqueryTargetlist = NIL;
+	List *newInsertTargetlist = NIL;
+	int resno = 1;
+	Index insertTableId = 1;
+	int updatedSubqueryEntryCount = 0;
+	Oid insertRelationId = InvalidOid;
+	int subqueryTargetLength = 0;
+
+	/* we only apply the reording for INSERT ... SELECT queries */
+	if (!InsertSelectQuery(originalQuery))
+	{
+		return originalQuery;
+	}
```

Comment:
Personally I think it'd make a lot more sense to check this outside, rather than in this function. We need to know that at the callsite anyway, and this just overloads the function with multiple tasks.